### PR TITLE
[SPARK-23390][SQL] Flaky Test Suite: FileBasedDataSourceSuite in Spark 2.3/hadoop 2.7

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This test only fails with sbt on Hadoop 2.7, I can't reproduce it locally, but here is my speculation by looking at the code:
1. FileSystem.delete doesn't delete the directory entirely, somehow we can still open the file as a 0-length empty file.(just speculation)
2. ORC intentionally allow empty files, and the reader fails during reading without closing the file stream.

This PR improves the test to make sure all files are deleted and can't be opened.

## How was this patch tested?

N/A

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @sameeragarwal @dongjoon-hyun @gatorsmile

File: sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala [hunk 1]
```diff
@@ -102,17 +104,27 @@ class FileBasedDataSourceSuite extends QueryTest with SharedSQLContext {
...
+
           val df = spark.read.format(format).load(
             new Path(basePath, "first").toString,
             new Path(basePath, "second").toString,
             new Path(basePath, "third").toString)
 
-          val fs = thirdPath.getFileSystem(spark.sparkContext.hadoopConfiguration)
+          // Make sure all data files are deleted and can't be opened.
+          files.foreach(f => fs.delete(f, false))
           assert(fs.delete(thirdPath, true))
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala]: Hmmmm .. but it asserts true on the delete completion .. I would be surprised if it's something not guaranteed ..

**Comment by HyukjinKwon** [comment]
BTW, I would bet for case 2. in the PR description (just a rough wild guess).

**Comment by HyukjinKwon** [comment]
BTW, my rough wild guess was that case 2. (reading it but not closing it) happens in schema inference path.

**Comment by HyukjinKwon** [comment]
I won't get in the way but I am less sure on this. I thought this is also flaky in PR builder too anyway.

**Comment by cloud-fan** [comment]
> BTW, my rough wild guess was that case 2. (reading it but not closing it) happens in schema inference path.

According to the log, the leaked file stream was created when building the ORC columnar reader.

**Comment by HyukjinKwon** [comment]
You are right. I have run out of ideas. LGTM too for a try if it happens more frequently in spark-branch-2.3-test-sbt-hadoop-2.7.

**Comment by kiszk** [comment]
I am also thinking about this. I agree with this.
> According to the log, the leaked file stream was created when building the ORC columnar reader.

I am suspicious about relationship between `afterEach()` and `addTaskCompletionListener` (call `close()`). But, not sure. Let us try this approach first.

**Comment by sameeragarwal** [comment]
merging this to master/2.3. Thanks!

**Comment by cloud-fan** [comment]
> I am suspicious about relationship between afterEach() and addTaskCompletionListener (call close()). But, not sure. Let us try this approach first.

This is one of my speculations. There 2 possibilities I can think of: 1) the task completion listener is not called before `afterEach`. 2) the orc columnar reader's `close` method doesn't close the file stream.

For 1), seems we've fixed it in https://github.com/apache/spark/commit/c5a31d160f47ba51bb9f8a4f3141851034640fc7 . For 2), I'm not sure and may need help from ORC folks.

**Comment by HyukjinKwon** [comment]
I think I rushed to take a look at the first time. Thanks for fixing this.

**Comment by gatorsmile** [comment]
Great! https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-branch-2.3-test-sbt-hadoop-2.7/ becomes green again!!!

**Comment by dongjoon-hyun** [comment]
My bad. Thank you, guys. For the following, I'll investigate it.
> According to the log, the leaked file stream was created when building the ORC columnar reader.

**Comment by dongjoon-hyun** [comment]
For the following case, I'll make a PR for Spark ORC columnar reader very soon.
> 2) the orc columnar reader's close method doesn't close the file stream.

**Comment by dongjoon-hyun** [comment]
I created a PR, https://github.com/apache/spark/pull/20590 .

**Pull request closed without merging**